doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1522,144 +1508,108 @@ Managing the local context
+   The second form adds a new hypothesis with a name chosen by Coq
+   or specified by :n:`@as_ipat`.  If :n:`@ltac_expr3` is specified, it is
+   applied to the new subgoal.
 
-      .. exn:: Proof is not complete.
-         :name: Proof is not complete. (assert)
-         :undocumented:
+   :n:`@as_ipat`
+     Gives the name for the hypothesis or specifies how to split it into parts.
+     See :ref:`intropatterns`.
 
-   .. tacv:: assert @type as @simple_intropattern
+   :n:`by @ltac_expr3`
+     Applies :n:`@ltac_expr3` to solve the subgoal generated by :n:`assert`.
 
-      If :n:`simple_intropattern` is an intro pattern (see :ref:`intropatterns`),
-      the hypothesis is named after this introduction pattern (in particular, if
-      :n:`simple_intropattern` is :n:`@ident`, the tactic behaves like
-      :n:`assert (@ident : @type)`). If :n:`simple_intropattern` is an action
-      introduction pattern, the tactic behaves like :n:`assert @type` followed by
-      the action done by this introduction pattern.
+     If not specified, the tactic applies `exact @term` by default where
+     :token:`type` is the type of :token:`term` and also equivalent to
+     :tacn:`pose proof`.  If the head of term is :token:`ident`, the tactic
+     is equivalent to :tacn:`specialize`.
 
-   .. tacv:: assert @type as @simple_intropattern by @tactic
+   .. exn:: Not a proposition or a type.
 
-      This combines the two previous variants of :tacn:`assert`.
+      Arises when the argument :token:`type` is neither of type :g:`Prop`,
+      :g:`Set` nor :g:`Type`.
 
-   .. tacv:: assert (@ident := @term)
+   .. exn:: Proof is not complete.
+      :name: Proof is not complete. (assert)
 
-      This behaves as :n:`assert (@ident : @type) by exact @term` where
-      :token:`type` is the type of :token:`term`. This is equivalent to using
-      :tacn:`pose proof`. If the head of term is :token:`ident`, the tactic
-      behaves as :tacn:`specialize`.
+      :n:`@ltac_expr3` was not able to prove the provided :n:`@term` or :n:`@one_term`.
 
-      .. exn:: Variable @ident is already declared.
-         :undocumented:
+   .. exn:: Variable @ident is already declared.
+      :undocumented:
 
-.. tacv:: eassert @type as @simple_intropattern by @tactic
-   :name: eassert
+.. tacn:: eassert ( @ident : @term ) {? by @ltac_expr3 }
+          eassert @one_term {? @as_ipat } {? by @ltac_expr3 }
+   :name: eassert; _
 
-   While the different variants of :tacn:`assert` expect that no existential
-   variables are generated by the tactic, :tacn:`eassert` removes this constraint.
-   This lets you avoid specifying the asserted statement completely before starting
-   to prove it.
+   While the different variants of :tacn:`assert` expect that no :term:`existential
+   variables <existential variable>` are generated by the tactic, :tacn:`eassert`
+   removes this constraint.  This lets you avoid specifying the asserted statement
+   completely before starting to prove it.
```

Comment:
Concretely, this means that the statement may contain holes (`_`).

Reply:
We should talk about holes in general.  IIUC they never appear in the context.  Each hole appearing in `eassert` becomes a new existential variable?  They can also be used for several other purposes (intropatterns, match expressions, lets, function parameters, hints paths, universe levels) and my impression is that implicit arguments (still not well documented) will fill in some of the holes.

Reply:
For `eassert`, if the term has a `match` with `_` in the pattern, that's not going to be preserved rather than becoming an existential variable.  Right?  Can the condition for replacement be stated concisely?

Reply:
> Each hole appearing in `eassert` becomes a new existential variable? They can also be used for several other purposes (intropatterns, match expressions, lets, function parameters, hints paths, universe levels) and my impression is that implicit arguments (still not well documented) will fill in some of the holes.

You are right on all of this AFAICT.

> For `eassert`, if the term has a `match` with `_` in the pattern, that's going to be preserved rather than becoming an existential variable, right? Can the condition for replacement be stated concisely?

Actually, `match` with `_` is always elaborated to not use any `_` anymore if I'm not mistaken. But this is indeed, not a hole that creates an existential variable. Holes that correspond to existential variables are the ones which appear in position where a term is expected (not a pattern).